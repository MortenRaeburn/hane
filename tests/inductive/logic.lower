Inductive False : Prop :=.
Inductive True : Prop :=
    | I : True.
Inductive and (Prop) (Prop) : Prop :=
    | conj : forall['1] (forall['1] (((and) ('3)) ('2))).
Inductive or (Prop) (Prop) : Prop :=
    | or_introl : forall['1] (((or) ('2)) ('1))
    | or_intror : forall['0] (((or) ('2)) ('1)).
Inductive ex (Set) (forall['0] (Prop)) : Prop :=
    | ex_intro : forall['1] (forall[('1) ('0)] (((ex) ('3)) ('2))).
Inductive eq (Set) ('0) : forall['1] (Prop) :=
    | eq_refl : (((eq) ('1)) ('0)) ('0).
