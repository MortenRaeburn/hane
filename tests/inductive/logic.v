Inductive False : Prop :=.
Inductive True: Prop :=
    I: True.
Inductive and (P : Prop) (Q : Prop) : Prop :=
    conj : forall (p : P) (q : Q), and P Q.
Inductive or (P : Prop) (Q : Prop) : Prop :=
    | or_introl : forall p : P, or P Q
    | or_intror : forall q : Q, or P Q.

Inductive ex (T : Set) (P : forall x : T, Prop) : Prop :=
    ex_intro : forall (x : T) (p : P x), ex T P.

Inductive eq (T : Set) (x : T) : forall y : T, Prop :=
    eq_refl : eq T x x.

Definition ex_falso : forall (P : Prop) (f : False), P :=
    fun (P : Prop) (f : False) =>
        match f as f in False return P with end.

Definition and_elem : forall (P : Prop) (Q : Prop) (R : Prop) (H : forall (p : P) (q : Q), R) (p : and P Q), R :=
    fun (P : Prop) (Q : Prop) (R : Prop) (H : forall (p : P) (q : Q), R) (p : and P Q) =>
        match p as p in and _ _ return R with
        | conj _ _ p q => H p q
        end.

Definition or_elem : forall (P : Prop) (Q : Prop) (R : Prop) (HP : forall p : P, R) (HQ : forall q : Q, R) (o : or P Q), R :=
    fun (P : Prop) (Q : Prop) (R : Prop) (HP : forall p : P, R) (HQ : forall q : Q, R) (o : or P Q) =>
        match o as o in or _ _ return R with
        | or_introl _ _ p => HP p
        | or_intror _ _ q => HQ q
        end.

Definition ex_elem : forall (T : Set) (P : forall x : T, Prop) (R : Prop) (H : forall (x : T) (p : P x), R) (e : ex T P), R :=
    fun (T : Set) (P : forall x : T, Prop) (R : Prop) (H : forall (x : T) (p : P x), R) (e : ex T P) =>
        match e as e in ex _ _ return R with
        | ex_intro _ _ x p => H x p
        end.

Definition eq_elem : forall (T : Set) (x : T) (R : forall x : T, Prop) (r : R x) (y : T) (e : eq T x y), R y :=
    fun (T : Set) (x : T) (R : forall x : T, Prop) (r : R x) (y : T) (e : eq T x y) =>
        match e as e in eq _ _ z return R z with
        | eq_refl _ _ => r
        end.

Definition eq_sym : forall (T : Set) (x : T) (y : T) (e : eq T x y), eq T y x :=
    fun (T : Set) (x : T) (y : T) (e : eq T x y) =>
        match e as e in eq _ _ z return eq T z x with
        | eq_refl _ _ => eq_refl T x
        end.

Definition eq_trans : forall (T : Set) (x : T) (y : T) (z : T) (e1 : eq T x y) (e2 : eq T y z), eq T x z :=
    fun (T : Set) (x : T) (y : T) (z : T) (e1 : eq T x y) (e2 : eq T y z) =>
        match e2 as e2 in eq _ _ z return eq T x z with
        | eq_refl _ _ => e1
        end.

Definition f_equal : forall (S : Set) (T : Set) (f : forall x : S, T) (x : S) (y : S) (e : eq S x y), eq T (f x) (f y) :=
    fun (S : Set) (T : Set) (f : forall x : S, T) (x : S) (y : S) (e : eq S x y) =>
        match e as e in eq _ _ z return eq T (f x) (f z) with
        | eq_refl _ _ => eq_refl T (f x)
        end.
