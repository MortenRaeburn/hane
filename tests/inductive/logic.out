Inductive False : Prop :=.
Inductive True : Prop :=
    | I : True.
Inductive and (Prop) (Prop) : Prop :=
    | conj : forall['1] (forall['1] (((and) ('3)) ('2))).
Inductive or (Prop) (Prop) : Prop :=
    | or_introl : forall['1] (((or) ('2)) ('1))
    | or_intror : forall['0] (((or) ('2)) ('1)).
Inductive ex (Set) (forall['0] (Prop)) : Prop :=
    | ex_intro : forall['1] (forall[('1) ('0)] (((ex) ('3)) ('2))).
Inductive eq (Set) ('0) : forall['1] (Prop) :=
    | eq_refl : (((eq) ('1)) ('0)) ('0).
Definition ex_falso : forall[Prop] (forall[False] ('1)) := fun[Prop] (fun[False] (match '0 in False return '2 with end)).
Definition and_elem : forall[Prop] (forall[Prop] (forall[Prop] (forall[forall['2] (forall['2] ('2))] (forall[((and) ('3)) ('2)] ('2))))) := fun[Prop] (fun[Prop] (fun[Prop] (fun[forall['2] (forall['2] ('2))] (fun[((and) ('3)) ('2)] (match '0 in and return '5 with (('5) ('1)) ('0) end))))).
Definition or_elem : forall[Prop] (forall[Prop] (forall[Prop] (forall[forall['2] ('1)] (forall[forall['2] ('2)] (forall[((or) ('4)) ('3)] ('3)))))) := fun[Prop] (fun[Prop] (fun[Prop] (fun[forall['2] ('1)] (fun[forall['2] ('2)] (fun[((or) ('4)) ('3)] (match '0 in or return '6 with ('5) ('0) | ('4) ('0) end)))))).
Definition ex_elem : forall[Set] (forall[forall['0] (Prop)] (forall[Prop] (forall[forall['2] (forall[('2) ('0)] ('2))] (forall[((ex) ('3)) ('2)] ('2))))) := fun[Set] (fun[forall['0] (Prop)] (fun[Prop] (fun[forall['2] (forall[('2) ('0)] ('2))] (fun[((ex) ('3)) ('2)] (match '0 in ex return '5 with (('5) ('1)) ('0) end))))).
Definition eq_elem : forall[Set] (forall['0] (forall[forall['1] (Prop)] (forall[('0) ('1)] (forall['3] (forall[(((eq) ('4)) ('3)) ('0)] (('3) ('1))))))) := fun[Set] (fun['0] (fun[forall['1] (Prop)] (fun[('0) ('1)] (fun['3] (fun[(((eq) ('4)) ('3)) ('0)] (match '0 in eq return ('7) ('1) with '4 end)))))).
Definition eq_sym : forall[Set] (forall['0] (forall['1] (forall[(((eq) ('2)) ('1)) ('0)] ((((eq) ('3)) ('1)) ('2))))) := fun[Set] (fun['0] (fun['1] (fun[(((eq) ('2)) ('1)) ('0)] (match '0 in eq return (((eq) ('7)) ('1)) ('6) with ((eq_refl) ('5)) ('4) end)))).
Definition eq_trans : forall[Set] (forall['0] (forall['1] (forall['2] (forall[(((eq) ('3)) ('2)) ('1)] (forall[(((eq) ('4)) ('2)) ('1)] ((((eq) ('5)) ('4)) ('2))))))) := fun[Set] (fun['0] (fun['1] (fun['2] (fun[(((eq) ('3)) ('2)) ('1)] (fun[(((eq) ('4)) ('2)) ('1)] (match '0 in eq return (((eq) ('9)) ('8)) ('1) with '3 end)))))).
Definition f_equal : forall[Set] (forall[Set] (forall[forall['1] ('1)] (forall['2] (forall['3] (forall[(((eq) ('4)) ('1)) ('0)] ((((eq) ('4)) (('3) ('2))) (('3) ('1)))))))) := fun[Set] (fun[Set] (fun[forall['1] ('1)] (fun['2] (fun['3] (fun[(((eq) ('4)) ('1)) ('0)] (match '0 in eq return (((eq) ('8)) (('7) ('6))) (('7) ('1)) with ((eq_refl) ('6)) (('5) ('4)) end)))))).
